### ISSUE TITLE
Stopped workspace has its deployment config at place and its pod running

### ISSUE THREAD

**Issue by agajdosi** [eclipse-che/che#12569]
### Description

When running tests of Che addon on Minishift I have noticed the workspace's deployment config and pod are preserved after the `stop` of runtime and `deletion` of workspace. When I went through logs on Che pod I have noticed the `stop` command did not run successfully, however even though the error the reported state is then `STOPPED`, without any troubles reported in workspace JSON (`"stoppedAbnormally":"false"`):

```
12:55:15 2019-01-31 12:54:46,151[aceSharedPool-1]  [ERROR] [o.e.c.a.w.s.WorkspaceRuntimes 536]   - Error occurred during stopping of runtime 'workspacemiff1794hzwtblm3:default:che' by user 'che'. Error: Error(s) occurs while cleaning up the namespace. Operation: [patch]  for kind: [Deployment]  with name: [workspacemiff1794hzwtblm3.dockerimage]  in namespace: [mini-che]  failed.
12:55:15 org.eclipse.che.api.workspace.server.spi.InfrastructureException: Error(s) occurs while cleaning up the namespace. Operation: [patch]  for kind: [Deployment]  with name: [workspacemiff1794hzwtblm3.dockerimage]  in namespace: [mini-che]  failed.
```

Not deleted Deployment (is it deployment config?) then redeploys the workspace pod and so it uses the resources even when STOPPED or Deleted. Seems to me that the `stop` should handle this case and react accordingly (is it possible to make it more robust or implement hard/force delete in case of error?). 

### Reproduction Steps

- minishift start
- minishift addons apply che
- create workspace
- run some commands
- stop workspace
- delete workspace

NOTE: it is reproducible for me only on weaker machines in CI

**OS and version:**    

CentOS, OpenShift 3.11.0, Che 6.17.1

**Diagnostics:** 
```
12:33:23 2019-01-30 12:20:35,731[nio-8080-exec-5]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 473]   - Workspace 'che/vert.x-stack-test' with id 'workspacen79p9px0dkcauptp' is stopping by user 'che'
12:33:23 2019-01-30 12:20:46,801[aceSharedPool-1]  [ERROR] [o.e.c.a.w.s.WorkspaceRuntimes 536]   - Error occurred during stopping of runtime 'workspacen79p9px0dkcauptp:default:che' by user 'che'. Error: Error(s) occurs while cleaning up the namespace. Operation: [patch]  for kind: [Deployment]  with name: [workspacen79p9px0dkcauptp.dockerimage]  in namespace: [mini-che]  failed.
12:33:23 org.eclipse.che.api.workspace.server.spi.InfrastructureException: Error(s) occurs while cleaning up the namespace. Operation: [patch]  for kind: [Deployment]  with name: [workspacen79p9px0dkcauptp.dockerimage]  in namespace: [mini-che]  failed.
12:33:23 	at org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace.doRemove(KubernetesNamespace.java:192)
12:33:23 	at org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProject.cleanUp(OpenShiftProject.java:100)
12:33:23 	at org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInternalRuntime.internalStop(KubernetesInternalRuntime.java:619)
12:33:23 	at org.eclipse.che.api.workspace.server.spi.InternalRuntime.stop(InternalRuntime.java:177)
12:33:23 	at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StopRuntimeTask.run(WorkspaceRuntimes.java:506)
12:33:23 	at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:38)
12:33:23 	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1626)
12:33:23 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
12:33:23 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
12:33:23 	at java.lang.Thread.run(Thread.java:748)
12:33:23 2019-01-30 12:20:51,497[nio-8080-exec-6]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 274]    - Workspace 'workspacen79p9px0dkcauptp' removed by user 'che'
```

However the workspace pod is still running, probably due to fact that deployment config has redeployed it:

```
12:55:44 NAME                                                    READY     STATUS    RESTARTS   AGE
12:55:44 che-1-xfrfl                                             1/1       Running   0          44m
12:55:44 workspacen79p9px0dkcauptp.dockerimage-bb48dcd64-6k5gc   1/1       Running   0     
```

**Comment by sleshchenko** [comment]
@amisevsk Maybe you have an idea why such exception might happen during Deployments removing.

**Comment by amisevsk** [comment]
I have no idea why this would happen. We've had strange issues where the deletion of the deployment did not cascade, leaving a pod still running, but never the complete failure of deletion.

If it's only reproducible on weak machines, perhaps it is a minishift / oc issue -- I've had minishift randomly start timing out on all requests in the past.

**Comment by amisevsk** [comment]
@agajdosi There are a few issues with this: 

1. `oc delete deployment <deployment>` is basically what Che does right now, except through the REST API. If it's failing that is a significant issue, and if it's only failing in certain limited scenarios fixing how that is handled is less of a priority.
   - FWIW from a user's perspective: failure to delete a deployment would basically mean that the next launch would fail (with error message about resources already existing). This failure would trigger a second cleanup which should remove the resources successfully (unless deletion through the API fails consistently, in which case that's a system requirements issue).
2. I agree that there are a number of failure modes that are not handled especially well by Che -- failure to clean up PVCs just leaves files on the PVC that have to be manually removed, workspace operations that fail generally just lead to "STOPPED" workspaces, but fixing this issue could be very complex. If a workspace is `BROKEN`, how is it broken? How do we store what needs to be done to resolve the issue, given that infra-wise, it could be any of a number of things? This is compounded in the multi-user case, as Che could require a user's token to actually perform the necessary actions, meaning that if Che is restarted, it might not even be able to resolve these issues. 
   - For instance, with workspace stopping: suppose a workspace stop fails. This could mean a) that the deployment sticks around (as in your case) or potentially b) the command timed out but succeeded (I believe I've seen this occur). In the case of a) the correct thing to do would be to retry the stop later, but in the case of b) retrying the stop would fail every time, since those resources don't exist. So if b) occurs and we don't detect the underlying issue, we're in trouble.
   - Additionally, suppose we want to retry the delete 15 minutes later. What if in the mean time, the user has restarted their workspace? We would have to update the state for `BROKEN` workspaces whenever the user interacts with Che.

Basically, it would be a significant undertaking to cover these edge cases, and unless they're common enough to actually cause an issue, I personally don't consider it a very high priority. In an ideal world we'd have a periodic job that retries failed operations, but I don't see a very simple path towards implementing this.
